gap/semigroups/semifp.gi

```diff
@@ -636,3 +636,163 @@ function(M)
+        od;
+
+        # converts a character to the element it represents
+        chartoel := function(char)
+            local i;
+            for i in [1 .. Size(gens)] do
+                if char = String(gens[i])[1] then
+                    return gens[i];
+                fi;
+            od;
+            ErrorNoReturn(Concatenation("expected a free semigroup generator",
+                          " but found ", [char]));
+        end;
+
+        # i acts as a pointer to positions in the string.
+        product := "";
+        i := 1;
+        while i <= Size(word) do
+            # if there are no brackets the character is left as it is.
+            if not word[i] = '(' then
+                if product = "" then
+                  product := chartoel(word[i]);
+                else
+                  product := product * chartoel(word[i]);
+                fi;
+            else
+                lbracket := i;
+                rbracket := -1;
+                # tracks how 'deep' the position of i is in terms of nested
+                # brackets
+                nestcount := 0;
+                i := i + 1;
+                while i <= Size(word) do
+                    if word[i] = '(' then
+                        nestcount := nestcount + 1;
+                    elif word[i] = ')' then
+                        if nestcount = 0 then
+                            rbracket := i;
+                            break;
+                        else
+                            nestcount := nestcount - 1;
+                        fi;
+                    fi;
+                    i := i + 1;
+                od;
+                # if rbracket is not followed by ^ then the value inside the
+                # bracket is appended (recursion is used to remove any brackets
+                # in this value)
+                if rbracket = Size(word) or (not word[rbracket + 1] = '^') then
+                    if product = "" then
+                      product := RemoveBrackets(
+                                 word{[lbracket + 1 .. rbracket - 1]});
+                    else
+                      product := product *
+                                 RemoveBrackets(
+                                 word{[lbracket + 1 .. rbracket - 1]});
+                    fi;
+                # if rbracket is followed by ^ then the value inside the
+                # bracket is appended the given number of time
+                else
+                    i := i + 2;
+                    while i <= Size(word) do
+                        if word[i] in "0123456789" then
+                            i := i + 1;
+                        else
+                            break;
+                        fi;
+                    od;
+
+                    p := Int(word{[rbracket + 2 .. i - 1]});
+                    if p = 0 then
+                      ErrorNoReturn(Concatenation("expected ^ to be followed",
+                      " by a positive integer but found 0"));
```

Comment:
Why is disallowed here but allowed above? Are you intending to allow things like a^01?

Reply:
That was the idea yes.